core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java

```diff
@@ -165,4 +203,46 @@ public Result apply(AggregationNode parent, Captures captures, Context context)
                         .setPreGroupedSymbols(ImmutableList.of())
                         .build());
     }
+
+    private boolean shouldAddMarkDistinct(AggregationNode aggregationNode, Context context)
+    {
+        if (aggregationNode.getGroupingKeys().isEmpty()) {
+            // global distinct aggregation is computed using a single thread. MarkDistinct will help parallelize the execution.
+            return true;
+        }
+        double numberOfDistinctValues = context.getStatsProvider().getStats(aggregationNode).getOutputRowCount();
```

Comment:
Are we sure that relying on aggregation output row count is safe ?
For multiple grouping keys, this is likely to be an overestimate (assumption that there is no correlation between columns). For non-partitioned columns of partitioned tables in hive connector we are underestimating NDVs by taking a max of NDVs from all partitions. Also, we sample statistics from only 100 partitions.
All these problems exist in existing usages as well (ApplyPreferredTableWriterPartitioning in particular). However, it seems risky to increase dependence on this even after we know of cases where bad statistics for the above reasons have caused problems in existing usages.

Reply:
> Are we sure that relying on aggregation output row count is safe ?

I'm sure there are cases (over-estimation of very small NDV) where relying on stats here will cause the rule to choose the wrong approach, but that goes for most CBO rules, right?
I can also increase the thresholds/multipliers to count overestimation in but that has negative consequences if the estimation is correct.

Reply:
I restricted reliance on the NDV stats only for a single column group by to avoid issues with overestimation